### PR TITLE
Major Update: Addition of 'anidb3' & 'anidb4' mode

### PR DESCRIPTION
**Pull request by EndOfLine369**
@ZeroQI,
HAMA mirror of https://github.com/ZeroQI/Absolute-Series-Scanner/pull/183

### CODE REVIEW

File: Contents/Code/AniDB.py [hunk 1]
```diff
@@ -113,7 +113,7 @@ def GetMetadata(media, movie, error_log, source, AniDBid, TVDBid, AniDBMovieSets
   
   ### Build the list of anidbids for files present ####
   if source.startswith("tvdb") or source.startswith("anidb") and not movie and max(map(int, media.seasons.keys()))>1:  #multi anidbid required only for tvdb numbering
-    full_array  = [ anidbid for season in Dict(mappingList, 'TVDB') or [] for anidbid in Dict(mappingList, 'TVDB', season) if season and 'e' not in season and anidbid.isdigit() ]
+    full_array  = [ anidbid for season in Dict(mappingList, 'TVDB') or [] for anidbid in Dict(mappingList, 'TVDB', season) if season and season.startswith('s') and 'e' not in season and anidbid.isdigit() ]
```
Comment (EndOfLine369) [inline Contents/Code/AniDB.py]: EX: {'s1': {'12560': '0'}, 's0': {'13950': '0'}, '13950': (0, '')}

**Comment by EndOfLine369** [comment]
Should now be set. Bar any issues from you Zero 😄

**Comment by ZeroQI** [comment]
Read each code individually, see no issues, should not impact other series which scared me
Tried to make hama code simple to allow people to learn and re-use and short vertically which I cannot follow otherwise... Seem like you did great in short time... How did you find the source code if I may ask?

**Comment by EndOfLine369** [comment]
> How did you find the source code if I may ask?

Took some time to wrap my brain around why the Dict & SaveDict existed. To return a default if does not exist yet & create parent dicts if does not already exist respectively.

Also took some time to figure out the format of the mappingList['TVDB'] dict.
I still do not know why/where they are added and what use those anidbid entries are for.
EX: `'13950': (0, '')` in `{'s1': {'12560': '0'}, 's0': {'13950': '0'}, '13950': (0, '')}`

The biggest annoyance in all of this was having to split everything up into its own GetMetadata and then only once all the pieces are pulled can I make the adjustments. A lot easier in the scanner as its all in one piece.

The 'dict_TheTVDB' was a simple season # adjustment. Once I got into my head the correct logic for it. Rewrote it here https://github.com/ZeroQI/Hama.bundle/commit/17da2d1b1e537596f28a1e656d34c25bc8a03128 and seems to now be good. Fingers crossed. But the 'dict_AniDB' had to be regenerated as you don't load in special info. But even if you did, finding what needs to be adjusted for each episode would be a lot more complicated then just regenerating it and trusting the existing GetMetadata logic. Also the reason I don't try to pull out special info from 'dict_TheTVDB' season 0 into the new locations. They just have to rely on the AniDB info alone.

**Comment by ZeroQI** [comment]
The Dict and saveDict allowed to save so much code by not testing if key existed for each metadata, and even append to lists without checking if it existed beforehands...
Am really proud of them...

So each metadata per source is saved per numbering used on disk so TheTVDB need to be saved in anidb numbering for single seasons. It needs to be able to find the anidb numbering from TheTVDB one so the mapping is saved to do the reverse conversion as the XML was done to do it anidb to TheTVDB only. Nobody to my knowledge cored it so had to invent a structure and I like dict  quite a bit...

Format: (to be edited once I get to work)

**Comment by EndOfLine369** [comment]
I understood the format already and matches from what you have stated. But the unknown is the `'13950': (0, '')` entries as I mentioned and are not in your format breakdown.

Found your actually adding these in from AniDB.py. Was this by accident as you meant to store it somewhere other than mappingList['TVDB'] dict? Sounds like this should be stored somewhere else as it not related to the 'TVDB' mapping.
```python
        AniDBid_table = Dict(mappingList, 'TVDB') or {}
        for relatedAnime in xml.xpath('/anime/relatedanime/anime'):
          if relatedAnime.get('id') not in AniDBid_table:  AniDBid_table[relatedAnime.get('id')] = (0, "")
```

**Comment by ZeroQI** [comment]
ah don't think that is voluntary indeed, should have made a deep copy of the dict possibly...
the restrictions are to avoid other fields like defaulttvdb season and only get seasons and not s0e1 type keys neither

**Pull request merged**